docs/web/about/roadmap.md

```diff
@@ -2,72 +2,85 @@
+## Make SQL and SQLAlchemy base backends independent
+
+At the moment, Ibis has two different ways of implementing SQL-based backends,
+with SQLAlchemy, and with a custom SQL compiler. For historical reasons, the
+SQLAlchemy backend was built subclassing the custom Ibis SQL compiler. But
+conceptually, this doesn't make sense. The right way to have the code structured
+would be to have a base backend with the code shared among both backends (and
+possibly other non-SQL backends). And then have both base backends subclass that.
+
+## Move SQL base backend, and backends based on it to external repositories
+
+Ideally, for SQL backends, Ibis should provide one single way to generate the
+SQL. Using SQLAlchemy. For historical reasons, Ibis implemented its custom
+SQL generator/compiler. But this implementation is basically a suboptimal copy
+of SQLAlchemy, and it doesn't add value, but it does add a lot of code to
+maintain and complexity. In the long-term, the ideal status would be to have
+all SQL backends implemented based on SQLAlchemy. It will take some time
+to have the current Ibis compiler based backends moved to SQLAlchemy. But to
+make this transition easy, would be good to remove these backends (including
+the base SQL backend) in separate projects. So, new backends based on SQLAlchemy
+can be developed in parallel, without conflicts or extra complexity, and
+users can decide which one to use for a given backend.
+
+## Simplify development of SQLAlchemy backends
+
+Given a SQLAlchemy engine, Ibis should be able to have most of the information
+it needs to work with that engine. There are cases where backends need to
+extend SQLAlchemy engines, such as with non-standard operations, custom types,
+UDFs or GIS functionality. Other than those cases, for simple queries, Ibis
+should be able to simply work given a SQLAlchemy engine. We should make Ibis
+work with arbitrary engines without implementing a backend. And for engines
+that are extended, it should be possible to write backends in the simplest and
+fastest way.
+
+## Execution output
+
+Currently, the output of executing an expression (i.e. `expr.execute()`) is
+returned as a pandas DataFrame. This is convenient, and the output can
+be further processed with the extensive features of pandas, or saved to
+different formats, with the pandas exporting interfaces. But in some cases,
+exporting to pandas implies adding pandas as an undesired dependency, or
+having an extra overheat. Often, Ibis backends are implemented using a
+[Python DB API 2.0](https://www.python.org/dev/peps/pep-0249/) interface,
+and accessing its Cursor object can be more convenient. In other cases,
+serializing from the backend to Apache Arrow is possible (and efficient)
+and accessing in this format would be the preferred way. Ibis should provide
+a more flexible way to access the results. For example, via an Ibis option,
+an argument (e.g. `.execute(format='cursor')`), different methods for
```

Comment:
Hmmm... This could work! A pyarrow table could be another option. Maybe even "fletcher" if a pandas dataframe backed by arrow is desired?

Reply:
I think there are variety of ways to handle the conversion problem. IMO a string argument is not desirable here due to its inflexible nature and lack of type information.

How about an API that allows anyone to implement custom format conversion while preserving backwards compatibility? Here's one way this could look:

#### `.execute(converter=MyConverter())`

This API is flexible enough to allow any existing conversion plus any custom conversion that someone may want to implement.

**Benefits**

* Tons of flexibility. If someone doesn't like the way things are being done in ibis, they can just write their own subclass and not have to patch the codebase.
* Existing behavior of execute is preserved, there'd be an `ibis.converters.pandas` singleton that'd be the default value of `converter` for the `execute` method.
*  Maintainability. Ibis maintainers need only write and maintain the abstract interface and possibly a core set of converters. Everything else can live in a separate set of libraries.

**Drawbacks**

* Potential for misuse: folks may start to put any and every post processing step inside of converters. This is probably fine, since there's no reason to dictate what a custom subclass for conversion would look like. I think we'd just have to be careful internally to not turn converters into god classes.

Reply:
Sounds like a good idea. Using `format='cursor'` in the roadmap is just to make clear what this point means. Having something scalable sounds good, even if I'm not sure there are so many different formats. Maybe returning a DBAPI2 cursor that can easily be converted to Arrow or pandas can be a good idea. If we implement converters, we need to have an internal (or not) format that converters are going to be converted from. If it makes sense that this representation is for example a DBAPI2 cursor, returning it directly with `.execute()`, and then being able to do something like `expr.execute().convert(ibis.converters.arrow)` could also be an option.

In any case, as for the roadmap, if we agree this is something we want to have, I think it's enough for now. When we start work on this, we can discuss all the different options in more detail.

Reply:
Can you change the text so it doesn't spell out any particular APIs? Something along the lines of "we're going to make it easier to not be forced into a pandas thing" :sweat_smile: